openshift/machine-config-operator
# OCPBUGS-35800: kubelet config controller does not always propogate tlsSecurityProfile

**- What I did**

- All subcontrollers within the kubelet-config-controller will now apply the `APIServer` object settings, including their bootstrap variants. If a bootstrap `APIServer` manifest is provided, the controllers will use that - otherwise it will default to the intermediate TLS profile.
- All the controllers will resync on updates to the `APIServer` object. The initial resync is done by the node config controller, which subsequently calls the kubelet-config and feature sub controllers.
- Also added some unit tests to verify the above behavior.

**- How to verify it**

1. Bring up a cluster with an install time `APIServer` manifest named "cluster". Ensure that this manifest has [the TLS profile you'd like to test](https://docs.openshift.com/container-platform/4.15/security/tls-security-profiles.html#tls-profiles-kubernetes-configuring_tls-security-profiles). 
2. Once the install is complete, observe the kubelet configs(located at `/etc/kubernetes/kubelet.conf`) on any node to ensure that it has the correct values for TLSCiphers and TLSMinVersion.
3. Switch between TLS profiles by updating the `APIServer` object named "cluster".
4. This will cause a new `MachineConfig` rollout. You can verify the TLS values on the nodes are as expected as the update rolls through the cluster.
5. Now, try applying a user defined `KubeletConfig` with TLS settings that are different from the ones in the `APIServer` object. This should cause another `MachineConfig` rollout to the pool it was targeted to. After the update, you should see the nodes in the pool with the TLS values specified by the new `KubeletConfig`(not the `APIServer` object).

**Things to note**

- The APIServer rejects the Modern profile at the moment, according to the docs.
- If no tlsSecurityProfile is provided, the MCO will default to Intermediate.
